testplans/bitswap/main.go

```diff
@@ -0,0 +1,181 @@
+		runenv.RecordMessage("running provider")
+		err = runProvide(ctx, runenv, h, bstore, ex)
+	case "requestors":
+		runenv.RecordMessage("running requestor")
+		err = runRequest(ctx, runenv, h, bstore, ex)
+	default:
+		runenv.RecordMessage("not part of a group")
+		err = errors.New("unknown test group id")
+	}
+	return err
+}
+
+func runProvide(ctx context.Context, runenv *runtime.RunEnv, h host.Host, bstore blockstore.Blockstore, ex exchange.Interface) error {
+	tgc := sync.MustBoundClient(ctx, runenv)
+	ai := peer.AddrInfo{
+		ID:    h.ID(),
+		Addrs: h.Addrs(),
+	}
+	tgc.MustPublish(ctx, providerTopic, &ai)
+	tgc.MustSignalAndWait(ctx, readyState, runenv.TestInstanceCount)
+
+	size := runenv.SizeParam("size")
+	count := runenv.IntParam("count")
+	for i := 0; i <= count; i++ {
+		runenv.RecordMessage("generating %d-sized random block", size)
+		buf := make([]byte, size)
+		rand.Read(buf)
+		blk := block.NewBlock(buf)
+		err := bstore.Put(blk)
+		if err != nil {
+			return err
+		}
+		err = ex.HasBlock(blk)
+		if err != nil {
+			return err
+		}
+		mh := blk.Multihash()
+		runenv.RecordMessage("publishing block %s", mh.String())
+		tgc.MustPublish(ctx, blockTopic, &mh)
+	}
+	tgc.MustSignalAndWait(ctx, readyDLState, runenv.TestInstanceCount)
+	tgc.MustSignalAndWait(ctx, doneState, runenv.TestInstanceCount)
+	return nil
+}
+
+func runRequest(ctx context.Context, runenv *runtime.RunEnv, h host.Host, bstore blockstore.Blockstore, ex exchange.Interface) error {
+	tgc := sync.MustBoundClient(ctx, runenv)
+	providers := make(chan *peer.AddrInfo)
+	blkmhs := make(chan *multihash.Multihash)
+	providerSub, err := tgc.Subscribe(ctx, providerTopic, providers)
+	if err != nil {
+		return err
+	}
+	ai := <-providers
+	if err != nil {
+		return err
+	}
+
+	runenv.RecordMessage("connecting  to provider provider: %s", fmt.Sprint(*ai))
+	providerSub.Done()
+
+	err = h.Connect(ctx, *ai)
+	if err != nil {
+		return fmt.Errorf("could not connect to provider: %w", err)
+	}
+
+	runenv.RecordMessage("connected to provider")
+
+	blockmhSub, err := tgc.Subscribe(ctx, blockTopic, blkmhs)
+	if err != nil {
+		return fmt.Errorf("could not subscribe to block sub: %w", err)
+	}
+	defer blockmhSub.Done()
+
+	// tell the provider that we're ready for it to publish blocks
+	tgc.MustSignalAndWait(ctx, readyState, runenv.TestInstanceCount)
+	// wait until the provider is ready for us to start downloading
+	tgc.MustSignalAndWait(ctx, readyDLState, runenv.TestInstanceCount)
+
+	begin := time.Now()
+	count := runenv.IntParam("count")
+	for i := 0; i <= count; i++ {
+		mh := <-blkmhs
+		runenv.RecordMessage("downloading block %s", mh.String())
+		dlBegin := time.Now()
+		blk, err := ex.GetBlock(ctx, cid.NewCidV0(*mh))
+		if err != nil {
+			return fmt.Errorf("could not download get block %s: %w", mh.String(), err)
+		}
+		dlDuration := time.Since(dlBegin)
+		runenv.RecordMessage("downloaded block %s", blk.Cid().String())
+		runenv.RecordMessage("download time: %s", dlDuration)
```

Comment:
Statistical quantities should be declared explicitly and encoded somehow, probably JSON, so we can parse them automatically. Perhaps create a stats.go file that declares output statistics, like:

```
type SingleDownloadSpeed struct {
    Cid string
    SingleDownloadDuration int64 // in nanoseconds
}
```

The stats.go file has to be in its own package (outside of the testplan) so that analysis code can access it.